Improvements to `ChatPerplexity` Integration

jamesliounis opened the pull request:
TL;DR:
1. Add `enable_search_classifier` parameter 
2. Update model default to `sonar-pro` 
3. Modify example notebook to reflect best practices

1. Updates the Perplexity API wrapper (ChatPerplexity) by adding a new parameter, `enable_search_classifier`, with a default value of `True`. This enables the use of a classifier to determine if search is necessary for a given API call. Additionally, the default model name has been updated from the deprecated "llama-3.1-sonar-small-128k-online" to "sonar-pro", keeping the integration current with the Perplexity API. The change allows users to both modify this new flag and rely on sensible defaults while preserving the flexibility of passing any additional parameters via `model_kwargs`.

5. Updates example notebook with best practices around using Perplexity and correct links to documentation. 

Dependencies:
No additional dependencies are introduced with this change.

Twitter handle:
(@LiounisJames, @PPLXDevs)

Additional steps performed:
	•	Ran make format, make lint, and make test to ensure that the changes adhere to the code quality and testing guidelines.

Thank you for reviewing this contribution!

Commit:
deploy pplx wrapper to langchain

Commit:
deploy pplx tool

Commit:
add search classifier, make temperature 0.2, change model

Commit:
add best practices to nb

Commit:
add example tool calling turning off search component

ccurme reviewed (commented):
Thank you!

Are we still intending to add the tools to langchain-community? Should we remove those?

Commit:
remove tool implementation

Commit:
Remove perplexity.py

jamesliounis commented:
@ccurme Removed the tool integration. Thanks for pointing this out.

sydney-runkle committed:
Merge branch 'master' into jamesliounis/update-chat-perplexity-implementation

james-pplx commented:
@ccurme anything else I can do to help with this?

james-pplx commented on libs/partners/perplexity/langchain_perplexity/chat_models.py:
```diff
@@ -147,9 +147,9 @@ class StructuredOutput(BaseModel):
 
     client: Any = None  #: :meta private:
-    model: str = "llama-3.1-sonar-small-128k-online"
+    model: str = "sonar-pro"
```
I changed is the default model because the one that was there (`llama-3.1-sonar-small-128k-online`) has been deprecated, we do not support it anymore. `sonar-pro` is our best performing model for a regular web search so it will most likely result in higher quality outputs, but if cost is a concern then we can use our cheapest alternative which is `sonar`. These are all of our currently supported [models](https://docs.perplexity.ai/models/model-cards).

ccurme replied:
IMO it's preferable to have the following:
- `model` has no default, users must pass it in on init
- `temperature` defaults to null and is only included in request payloads if it is explicitly specified.

That is, we mirror usage of the API.

I don't feel strongly about this if you'd prefer to maintain defaults in the LangChain integration. In either case, if we are changing the defaults we probably should bump the minor version from 0.1 to 0.2, though.

james-pplx commented on libs/partners/perplexity/langchain_perplexity/chat_models.py:
```diff
@@ -168,6 +168,8 @@ class StructuredOutput(BaseModel):
     """Whether to stream the results or not."""
     max_tokens: Optional[int] = None
     """Maximum number of tokens to generate."""
+    enable_search_classifier: bool = True
```
That's correct, it needs to be passed in the extra body. Forgot to push the commit. Fixing.

jamesliounis replied:
addressed these comments in this [commit](https://github.com/langchain-ai/langchain/pull/30802/commits/bb659f367289af5f0c5376bb2267d5b098aceef3). I removed the `enable_search_classifier` param, it's better to add it to the Perplexity documentation directly and let users pass it themselves if needed.

jamesliounis committed:
sonar-pro --> sonar + move enable_search_classifier to extra body

ccurme commented on libs/partners/perplexity/langchain_perplexity/chat_models.py:
```diff
@@ -39,8 +50,8 @@
 _BM = TypeVar("_BM", bound=BaseModel)
-_DictOrPydanticClass = Union[dict[str, Any], type[_BM], type]
-_DictOrPydantic = Union[dict, _BM]
+_DictOrPydanticClass = Union[Dict[str, Any], Type[_BM], Type]
```
Why are we un-doing https://github.com/langchain-ai/langchain/pull/30781?

ccurme commented on libs/partners/perplexity/langchain_perplexity/chat_models.py:
```diff
@@ -221,16 +232,17 @@ def validate_environment(self) -> Self:
             "max_tokens": self.max_tokens,
             "stream": self.streaming,
             "temperature": self.temperature,
+            "enable_search_classifier": self.enable_search_classifier,
```
I believe this is now broken.

jamesliounis replied:
Fixed. Removed it.

ccurme commented on docs/docs/integrations/chat/perplexity.ipynb:
```diff
@@ -295,51 +361,115 @@
+    "    decide whether to execute a full web search or bypass it, depending on the query.\n",
+    "    This helps to reduce unnecessary searches while still delivering accurate responses.\n",
+    "    \"\"\"\n",
+    "    llm = ChatPerplexity(model=\"sonar\", enable_search_classifier=True)\n",
```
This appears out of date.

ccurme reviewed (commented):
Would like to get this in but there are still some issues:

1. Code does not run. Please try running tests locally:
    - `make test` will run unit tests
    - `PPLX_API_KEY=pplx-... make integration_test` will run integration tests (with real HTTP requests)
2. We are (inadvertently?) undoing multiple PRs here (see comments)
    - Run `make format` and `make lint` locally to run the auto-formatter and linter, which will catch some of the errors.
4. Docs need to be updated to reflect most recent changes.

jamesliounis committed:
remove `enable_search_classifier` from the response

jamesliounis committed:
address latest comments

feat(chat): bump ChatPerplexity to v0.2, require model on init, conditional params

- Remove default for `model` and add a pre-init validator that errors (with model‑cards URL) if it’s not provided
- Make `temperature`, `max_tokens`, `stream`, and `timeout` only appear in the request payload when explicitly set
- Bump class docstring and version reference to v0.2

jamesliounis committed:
restore undone PRs

chore(chat_models): restore `_DictOrPydantic` alias and fix citations handling in `_generate`

- Re‑introduce the `_DictOrPydantic` type alias alongside `_DictOrPydanticClass`  
- In `_generate`, initialize `additional_kwargs` with `{"citations": response.citations}` and always append `images` and `related_questions`

jamesliounis committed:
Update README with supported model

jamesliounis committed:
Update perplexity.ipynb

jamesliounis committed:
add `enable_search_classifier` example

jamesliounis committed:
fix invalid json

jamesliounis commented:
I restored the undone PRs (which occurred because branch is now outdated) and have ran all the tests. 



<img width="1708" alt="Screenshot 2025-04-22 at 3 25 00 PM" src="https://github.com/user-attachments/assets/c60a3cbe-37e7-458a-a96f-4347eb8007cf" />
<img width="1713" alt="Screenshot 2025-04-22 at 3 25 08 PM" src="https://github.com/user-attachments/assets/1d646d8a-ff49-4e52-b992-9e2512523e6b" />
<img width="1714" alt="Screenshot 2025-04-22 at 3 25 16 PM" src="https://github.com/user-attachments/assets/ca5e70a2-2fcb-489b-adf2-0ed6d10d79e9" />

jamesliounis committed:
Merge branch 'master' into jamesliounis/update-chat-perplexity-implementation

ccurme reviewed (commented):
Still seeing tests failing on CI, undoing previous PRs and other unrelated changes.

ccurme commented:
Closing as author is unresponsive but please ping me if we want to keep iterating!
